src/jit/optcse.cpp

```diff
@@ -220,74 +202,125 @@ Compiler::fgWalkResult Compiler::optPropagateNonCSE(GenTree** pTree, fgWalkData*
     return WALK_CONTINUE;
 }
 
-/*****************************************************************************
- *
- *  Helper passed to Compiler::fgWalkAllTreesPre() to unmark nested CSE's.
- */
+//------------------------------------------------------------------------
+// Compiler::optUnmarkCSEs
+//    Helper passed to Compiler::fgWalkAllTreesPre() to unmark nested CSE's.
+// 
+//    argument 'pTree' is a pointer to a GenTree*
+//    argument 'data' contains pCallbackData which needs to be cast from void*
+//       to our actual 'wbKeepList' a pointer to a GenTree*
+//
+//    Any nodes visited that are in the 'keepList' are kept and we
+//    return WALK_SKIP_SUBTREES for them
+//    For any node not in the 'keepList' we call optUnmarkCSE and either
+//    remove the node and decrement the LclVar ref counts
+//    or add the subtree to 'wbKeepList' using gtBuildCommaList
+//    and return WALK_SKIP_SUBTREES.
+//
 
 /* static */
 Compiler::fgWalkResult Compiler::optUnmarkCSEs(GenTree** pTree, fgWalkData* data)
 {
-    GenTree*  tree     = *pTree;
-    Compiler* comp     = data->compiler;
-    GenTree*  keepList = (GenTree*)(data->pCallbackData);
+    GenTree*  tree       = *pTree;
+    Compiler* comp       = data->compiler;
+    GenTree** wbKeepList = (GenTree**)(data->pCallbackData);
+    GenTree*  keepList   = nullptr;
 
-    // We may have a non-NULL side effect list that is being kept
+    // If we have a non-Null wbKeepList
     //
-    if (keepList)
+    if (wbKeepList)
     {
-        GenTree* keptTree = keepList;
-        while (keptTree->OperGet() == GT_COMMA)
+        keepList = *wbKeepList;
+
+        // We may already have a non-NULL side effect list that is being kept
+        //
+        if (keepList)
         {
-            assert(keptTree->OperKind() & GTK_SMPOP);
-            GenTree* op1 = keptTree->gtOp.gtOp1;
-            GenTree* op2 = keptTree->gtGetOp2();
+            GenTree* keptTree = keepList;
 
-            // For the GT_COMMA case the op1 is part of the orginal CSE tree
-            // that is being kept because it contains some side-effect
-            //
-            if (tree == op1)
+            while (keptTree->OperGet() == GT_COMMA)
```

Comment:
Do we need to worry about quadratic behavior here?

Reply:
Not really, we can only move a node once from the original tree into the keepList.
The keepList uses GT_COMMA nodes to form a list of such expressions and we have to interate this list to locate the orginal nodes.

Reply:
I'm not worried  about the impact of adding, but the impact of searching. Since you are searching it seems like you expct to encounter the same tree multiple times. The first time you see it you will fail to find it via search and then add it. The other times you will find it via search and not add it.

So if there a N trees total that get added and each one is searched for M times, then we end up with M * N kind of perf impact.

What is it that keeps N and M small?